File: solutions/3625. Count Number of Trapezoids II/3625.py
import math
from collections import defaultdict, Counter

class Solution:
    def countTrapezoids(self, points: list[list[int]]) -> int:
        n = len(points)
        # Map slope (dx, dy) → list of segments (i, j)
        slope2segs = defaultdict(list)

        for i in range(n):
            x1, y1 = points[i]
            for j in range(i+1, n):
                x2, y2 = points[j]
                dx = x2 - x1
                dy = y2 - y1
                # normalize slope
                if dx == 0:
                    dx_norm, dy_norm = 0, 1
                elif dy == 0:
                    dx_norm, dy_norm = 1, 0
                else:
                    g = math.gcd(dx, dy)
                    dx_norm = dx // g
                    dy_norm = dy // g
                    # keep unique representation (force dx_norm > 0, or dx_norm==0 and dy_norm>0)
                    if dx_norm < 0:
                        dx_norm = -dx_norm
                        dy_norm = -dy_norm
                    elif dx_norm == 0 and dy_norm < 0:
                        dy_norm = -dy_norm
                slope2segs[(dx_norm, dy_norm)].append((i, j))

        total = 0
        # count all pairs of segments with same slope → potential trapezoids (parallel sides)
        for segs in slope2segs.values():
            m = len(segs)
            total += m * (m - 1) // 2

        # subtract parallelograms (which we counted twice)
        mid_slope_count = Counter()
        for slope, segs in slope2segs.items():
            for (i, j) in segs:
                x1, y1 = points[i]
                x2, y2 = points[j]
                mx = x1 + x2  # doubled midpoint x (to stay integer)
                my = y1 + y2  # doubled midpoint y
                mid_slope_count[(mx, my, slope)] += 1

        parallelograms = 0
        for cnt in mid_slope_count.values():
            if cnt >= 2:
                parallelograms += cnt * (cnt - 1) // 2

        return total - parallelograms

File: solutions/3625. Count Number of Trapezoids II/3625.cpp
class Solution {
public:
    int countTrapezoids(vector<vector<int>>& points) {
        int n = points.size();
        unordered_map<float,vector<float>> slope_intercept;
        unordered_map<int, vector<float>> midp_slope;
        for(int i=0;i<n;i++){
            int x1 = points[i][0], y1 = points[i][1];
            for(int j=i+1;j<n;j++){
                int x2 = points[j][0], y2 = points[j][1];
                int dy = y2 - y1, dx = x2 - x1;
                float slope,intercept;
                if(dx==0){
                    slope = 1e9 + 7;
                    intercept = x1;
                }else{
                    slope = (float)(dy)/(dx);
                    intercept = (float)(y1*dx - x1*dy)/(dx);
                }
                slope_intercept[slope].push_back(intercept);
                int midp = (x1 + x2)*10000 + (y1+y2);
                midp_slope[midp].push_back(slope);
            }
        }
        int ans = 0;
        for(auto &[slope, v_of_intercept] : slope_intercept){
            if(v_of_intercept.size()==0) continue;
            unordered_map<float,int> cnt;
            for(float intercept : v_of_intercept){
                cnt[intercept]++;
            }
            long long sum = 0, sum_of_sq = 0;
            for(auto &[k,v] : cnt){
                sum += v;
                sum_of_sq += (long long)v*v;
            }
            long long result = ((long long)sum*sum - sum_of_sq)/2;
            ans += result;
        }
        // subtracting parallelograms :
        for(auto &[midp, v_of_slope] : midp_slope){
            if(v_of_slope.size()==0) continue;
            unordered_map<float,int> cnt;
            for(float slope : v_of_slope){
                cnt[slope]++;
            }
            long long sum = 0, sum_of_sq = 0;
            for(auto &[k,v] : cnt){
                sum += v;
                sum_of_sq += (long long)v*v;
            }
            long long result = ((long long)sum*sum - sum_of_sq)/2;
            ans -= result;
        }
        return ans;
    }
};